### ISSUE TITLE
Implement Common Time Series Format

### ISSUE THREAD

**Issue by jranson** [tricksterproxy/trickster#348]
We currently have custom delta proxy cache code per time series origin, for doing things like merging, sorting, and cropping time series objects. Each origin type currently requires about 1000 lines of code to accomplish this, and it turns out that the code for each of the upstream origins is quite similar for these operations. I believe we can implement a common time series format that is capable of performing the time series operations generically. Then the only operations required of the origin type is the ability to marshal from the native json format into the common format struct, and vice versa. Doing so should reduce the code base to about 60% of its current size while adhering to good DRY principle.

**Comment by glinton** [comment]
Has there been any discussion/work done on this front? Sounds like a fantastic idea that would really streamline the adding of new origin types.

**Comment by jranson** [comment]
@glinton thanks for checking! So we have this feature slated for v1.2, which we are hoping to get out at the end of this quarter ([roadmap](https://github.com/tricksterproxy/trickster/blob/master/docs/roadmap.md))but it might push a few more weeks out. v1.1 will be cut in the next 36 hours and then we'll get to work on 1.2.

**Comment by jranson** [comment]
@glinton we are making good progress on the common format, and I should have that merged into the `v1.2.x` branch in the next few days. We've migrated prometheus over and it's working pretty flawlessly; I'm now working on porting InfluxDB 1.x and ClickHouse into the new format, and then we'll be ready to PR it all up. We've been able to reduce the memory footprint with the common format and reduce some expensive sorting, sizing and cloning routines that will make things even faster, so we can't wait to get the first beta out, hopefully in the next 10 days.

**Comment by jranson** [comment]
@glinton we've released Trickster 2.0 Beta 1, (previously called v1.2) with the Common Time Series Format, and tons of other new features. Sorry for the long delay!

One big nomenclature change is we now call origin types "providers". In this first beta, InfluxDB and other providers use custom un/marshalers to get from the provider's wire format into CTSF, and back. This was just because we started with ClickHouse (which does not provide a golang model pkg) and then used it as a template in the first cut of the remaining providers. In the next beta, we'll have re-implemented any available provider's models for unmarshaling (so long as they are performant), but will likely stick with custom marshalers to go directly from CTSF to provider wire format when responding downstream. 

So the implementor of a new provider would need to:
- Support the HTTP API of the provider
  - define http routes for the provider backend instances to the proxy, cache or timeseries delta cache engines
  - define what components of the HTTP request can be used to determine its caching profile (key hash methodology, time range extraction, etc)
  - manipulate any necessary components of the http request to modify the time range to only collect uncached data
- Convert from the provider's golang struct format to the CTSF (by iterating the data and transforming as needed)
- Iterate a CTSF object and output the JSON (or other supported content types) when responding on the wire (or optionally converting to the provider's model and mashaling if performant)

The good news is that an implementor no longer needs to worry about headbanging stuff like writing crop/merge/sort/etc. methods for multidimensional slices.
